autoscheduler/frontend/src/components/Schedule/Schedule.tsx

```diff
@@ -1,25 +1,252 @@
+  const schedule = useSelector<RootState, Meeting[]>((state) => state.meetings);
+  const availabilityList = useSelector<RootState, Availability[]>((state) => state.availability);
+  const availabilityMode = useSelector<RootState, AvailabilityType>(
+    (state) => state.availabilityMode,
+  );
+  const selectedAvailability = useSelector<RootState, AvailabilityArgs>(
+    (state) => state.selectedAvailability,
+  );
+  const dispatch = useDispatch();
+
+  /* state */
+  const [startDay, setStartDay] = React.useState(null);
+  // either the start or end time, in minutes since midnight,
+  // for the availability currently being added
+  const [time1, setTime1] = React.useState(null);
 
-  // helper functions for formatting
+  // state management for time display
+  const [hoveredDay, setHoveredDay] = React.useState(null);
+  const [mouseY, setMouseY] = React.useState<number>(null);
+  const [hoveredTime, setHoveredTime] = React.useState<number>(null);
+
+  // helper functions
   function formatHours(hours: number): number {
     return ((hours - 1) % 12) + 1;
   }
 
+  /**
+   * Given a MouseEvent in a calendar day, calculates the time, in minutes since midnight
+   * and rounded to the nearest 10, at which the mouse event was emitted
+   * @param evt
+   */
+  function eventToTime(evt: React.MouseEvent<HTMLDivElement, MouseEvent>): number {
+    const totalY = evt.currentTarget.clientHeight;
+    const yPercent = (evt.clientY - evt.currentTarget.getBoundingClientRect().top) / totalY;
+    const minutesPerDay = (LAST_HOUR - FIRST_HOUR) * 60;
+    const yMinutes = yPercent * minutesPerDay;
+    const roundedMinutes = Math.round(yMinutes / 10) * 10;
+    return roundedMinutes + FIRST_HOUR * 60;
+  }
+
+  /**
+   * Using time1 and time2 from the given availability, returns arguments for
+   * an availability that is at least 30 minutes long and ends before 9 PM
+   */
+  function roundUpAvailability(avl: AvailabilityArgs): AvailabilityArgs[] {
+    const blockSize = Math.abs(avl.time2 - avl.time1);
+    if (blockSize < 30) {
+      if (avl.time2 < 20 * 60 + 30) {
+        if (avl.time1 > 8 * 60 + 30) {
+          // if the availability is solidly between 8:30 and 20:30, just round up
+          return [{
+            ...avl,
+            // if the sign is zero, then assumes positive by default
+            time2: avl.time1 + 30 * (Math.sign(avl.time2 - avl.time1) || 1),
+          }];
+        }
+
+        // if availability is close to the eddges, force it to 8 to 8:30 in 2 steps
+        return [{
+          ...avl,
+          time2: 8 * 60,
+        }, {
+          ...avl,
+          time1: 8 * 60,
+          time2: 8 * 60 + 30,
+        }];
+      }
+      // new time blocks cannot be later than 9 PM / 2100
+      return [{
+        ...avl,
+        time2: 21 * 60,
+      }, {
+        ...avl,
+        time1: 21 * 60,
+        time2: 20 * 60 + 30,
+      }];
```

Comment:
The code that you tried is exactly what I had initially, but I had to change it to deal with a bug. The reason for this bug has to do with the implementation of `updateAvailability`. In particular, the way I determine which availability to update is by finding the availability currently in the Redux store whose start time or end time matches the provided `time1`. As such, if you try to replace the original availability with the 20:30 - 21:00 one in one go, the reducer will not find any existing availability that either starts or ends at 20:30, so the update will not be registered. 

The current implementation first replaces the end time of the existing availability with 21:00, so that in the second step, it can successfully find an availability whose start or end time is 21:00, and then update the start time of that availability to 20:30. Kinda confusing, I know, but it's the simplest way to get it to work without doing a ground-up rewrite of the availability reducer.